### PR TITLE
Standardize Windows service code

### PR DESCRIPTION
**Pull request by clarkb7**
### What does this PR do?

Removes the duplicated Windows service code and centralizes it in the `servicemain` package.

Removes the duplicated MESSAGETABLE code and centralizes it in the `messagestrings` package. Messages now have descriptive  names instead of hardcoded integers.

Removes the hard-fail if `eventlog.Open` fails at agent start. The event log messages are useful for debugging, but are not critical to agent functionality.

Replace `svc.IsAnInteractiveSession()` (deprecated) with `svc.IsWindowsService()`, so running agent commands via a non-interactive session (e.g. ssh) don't think its running as a Windows Service.

Re-enable `inv -e rtloader.format` in lint tests.

Fix incorrect usage of `fxutil.OneShot` by agent, system-probe, and dogstatsd. They each provided a `delayedFunc` that would start the agent and then return, which causes `fxutil.OneShot` to stop the fx app, triggering lifecycle `OnStop` hooks to be called, but then the agent keeps running. Each `delayedFunc` now blocks until the agent stops. This issue only manifested when running as a Windows Service because it uses a different code path than `agent run`.

### Motivation

Reduce duplicated code and ensure all agents have the same bug fixes
* Ensure very fast successful service stops delay exit so they don't break `Restart-Service`
* allow start when `eventlog.Open` fails
* Replace `svc.IsAnInteractiveSession()` (deprecated) with `svc.IsWindowsService()`
* run Windows service code before fx entry so errors can be caught, logged to event log, and reported to SCM

https://datadoghq.atlassian.net/browse/WA-163

https://datadoghq.atlassian.net/browse/WA-413

### Additional Notes

The service start semantics for each individual agent should be unchanged, it just uses shared code now.

The messagestrings package requires a C header to be generated before it will build. There is a new invoke task `inv build-messagetable` that does this. This has been bundled into `tasks/winbuildscripts/pre-go-build.ps1` along with the rtloader build that much of the agent requires.

### Describe how to test/QA your changes

Testing: MESSAGETABLE
- The messages are logged to the `Application` event log with the `Source` set to the service name. Start and stop each agent service successfully, and once more with an invalid yaml syntax in the configuration file, and ensure the start, stop and error are logged to the `Application` event log.

Testing: agent should start if eventlog.Open fails
- Run this command to restrict access to the Application event log `wevtutil sl Application /ca:'O:BAG:SYD:(A;;0xf0007;;;SY)(A;;0x7;;;BA)'`
- Confirm the change by examining the `channelAccess` field in the output of: `wevtutil gl Application`
- Ensure the agent services start

Testing: `Restart-Service` succeeds when disabling services
- enable process-agent, trace-agent, and system-probe in the agent configuration
- start all the services and ensure they are running
- disable process-agent, trace-agent, and system-probe in the agent configuration
- Run `Restart-Service -Force datadogagent` and ensure it succeeds without error

### Reviewer's Checklist

- [ ] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [ ] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] Changed code has automated tests for its functionality.
- [ ] Adequate QA/testing plan information is provided if the `qa/skip-qa` label is not applied.
- [ ] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the `k8s/<min-version>` label, indicating the lowest Kubernetes version compatible with this feature. 
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

### CODE REVIEW

File: releasenotes/notes/start-without-eventlog-access-44409fb7d9bb184b.yaml
Comment (drichards-87) [inline releasenotes/notes/start-without-eventlog-access-44409fb7d9bb184b.yaml:4]:
```suggestion
    The Agent now starts even if it cannot write events to the Application event log
```

**Review by drichards-87 — APPROVED** [review]
Left you one small suggestion from Docs.

File: tasks/winbuildscripts/lint.ps1
Comment (derekwbrown) [inline tasks/winbuildscripts/lint.ps1:11]: the root of vstudio is defined in an env variable; so if for some reason it moves this code wouldn't need to be changed.  IIRC it's `VSTUDIO_ROOT`

**Review by drichards-87 — APPROVED** [review]

**Review by just-chillin — APPROVED** [review]

File: cmd/process-agent/main_windows.go
Comment (just-chillin) [inline cmd/process-agent/main_windows.go:45]: Would it be possible to add some tests for the windows service behavior? It would be nice to have some code coverage in this area

**Review by gh123man — APPROVED** [review]
Dogstatsd + core agent startup + AML owned files LGTM.

**Review by knusbaum — COMMENTED** [review]
One question. Otherwise looks good.

File: pkg/util/winutil/servicemain/servicemain.go
Comment (iglendd) [inline pkg/util/winutil/servicemain/servicemain.go:147]: What is the point of that error? I mean from SCM reporting perspective. Would not it restarted the service because it exits without SCM service stopping? I saw few comments in other places but it is still not clear to me.

Previous code reported changes <- svc.Status{State: svc.Stopped} which seems to be intuitively at least a "right" thing to do.

File: pkg/util/winutil/servicemain/servicemain.go
Comment (iglendd) [inline pkg/util/winutil/servicemain/servicemain.go:134]: I wonder what SCM would report if you manipulate errno? I am asking if we can map some of our errors on SCM UI reporting error e.g.

File: pkg/util/winutil/servicemain/servicemain.go
Comment (iglendd) [inline pkg/util/winutil/servicemain/servicemain.go:185]: Should not exitGate to be initialized after
`changes <- svc.Status{State: svc.StartPending}`?

**Review by iglendd — COMMENTED** [review]
Much better than it was before

File: cmd/trace-agent/subcommands/run/command.go
Comment (clarkb7) [reply cmd/trace-agent/subcommands/run/command.go:27]: Yes, it's an fx thing. fx uses the most specific type available, which in this case is `context.cancelCtx` which isn't exported. I think I got it working with `fx.Provide` but since fx makes everything a singleton I didn't want to lock into one context.

File: cmd/agent/common/common_windows.go
Comment (clarkb7) [reply cmd/agent/common/common_windows.go:27]: [pkg/util/winutil/messagestrings/messagestrings.mc](https://github.com/DataDog/datadog-agent/pull/18702/files#diff-03b6be49a7f7e1147457b36e7adce2b7ca7617ed9613ff5d061c5f18eef0c284)

File: cmd/agent/subcommands/run/command.go
Comment (clarkb7) [reply cmd/agent/subcommands/run/command.go:231]: Yes, I think, when the agent becomes a component we can hopefully have just one `fxutil.OneShot`/`fxutil.Run` in this file.

File: cmd/process-agent/main.go
Comment (clarkb7) [reply cmd/process-agent/main.go:24]: its removal is defferred in [runAgent](https://github.com/DataDog/datadog-agent/pull/18702/files#diff-ceb61f735a2b8dbd182ee59e99b933a6fa931f3ea10c7730b287f47b7021a5b1L83-L88). `cleanupAndExitHandler` was necessary before because `runAgent` never returned, so the defer was never executed. But now that `runAgent` returns we can rely on the `defer` to cleanup the pid file.

File: pkg/util/winutil/servicemain/servicemain.go
Comment (clarkb7) [reply pkg/util/winutil/servicemain/servicemain.go:185]: it's meant to ensure the service is in the `RUNNING` state long enough to be considered "successfully started" by a service manager tool, so it needs to come after the service enters the `RUNNING` state. I'll clarify that in the comment.

File: pkg/util/winutil/servicemain/servicemain.go
Comment (iglendd) [reply pkg/util/winutil/servicemain/servicemain.go:161]: I think a bit summary at this code-line would be sufficient. I think I did understand your previous point in the code above but this code was out of the way and somewhat disjointed that connection did not occur to me.

**Review by iglendd — APPROVED** [review]
looks great

**Comment by clarkb7** [comment]
rebasing to fix snowflake build

File: cmd/system-probe/main_windows.go
Comment (brycekahle) [inline cmd/system-probe/main_windows.go:36]: Where does the context get sent to this channel? `StartSystemProbeWithDefaults` waits for a context to be sent.

File: cmd/system-probe/subcommands/run/command.go
Comment (brycekahle) [inline cmd/system-probe/subcommands/run/command.go:205]: Won't the error sent to `errChan` already be consumed by the above code? How would the caller use the returned channel?

File: cmd/system-probe/main_windows.go
Comment (clarkb7) [reply cmd/system-probe/main_windows.go:36]: oops, looks like I forgot to re-test after making the `fxutil.OneShot` change. will fix, thank you.

File: cmd/system-probe/subcommands/run/command.go
Comment (brycekahle) [inline cmd/system-probe/subcommands/run/command.go:198]: Can this channel be separate from the one returned to the caller? It was initially confusing to me on the order of operations. Since this one is only used within this function to know when `startSystemProbe` is done, it can be separate.

Comment (clarkb7) [reply cmd/system-probe/subcommands/run/command.go:198]: `fxutil.OneShot` can fail before calling the `delayedFunc` that calls `startSystemProbe`. I could separate them, but the commented on code would have to wait on both channels anyway. Would that be more clear?

Comment (brycekahle) [reply cmd/system-probe/subcommands/run/command.go:198]: So if `fxutil.OneShot` fails before `delayedFunc`, one value is sent to `errChan`. Otherwise `errChan` gets two values sent, correct (explicit send on 161 and return value from `fxutil.OneShot`).

In the latter case, the goroutine is blocked if an error OR if the caller doesn't consume the second value from the channel?

Comment (clarkb7) [reply cmd/system-probe/subcommands/run/command.go:198]: > So if fxutil.OneShot fails before delayedFunc, one value is sent to errChan. Otherwise errChan gets two values sent, correct? (explicit send on 161 and return value from fxutil.OneShot).

correct

> In the latter case, the goroutine is blocked if an error OR if the caller doesn't consume the second value from the channel?

I think the goroutine is blocked only in the case where`startSystemProbe` fails. The process will exit so it's not a big deal. But I think that can be fixed by moving the explicit send on 161 after the `if err != nil` check.

Comment (brycekahle) [reply cmd/system-probe/subcommands/run/command.go:198]: Let's do that. Excess goroutines can cause problems in tests.

**Review by brycekahle — APPROVED** [review]

File: pkg/util/winutil/servicemain/servicemain.go
Comment (clarkb7) [reply pkg/util/winutil/servicemain/servicemain.go:148]: if we set `dwWin32ExitCode` (`errno=1`), it interprets the error as a Windows Error.
image


if we set `dwServiceSpecificExitCode` (`errno=1`+`ssec=true`)
image
It's nice that it points the user to the event log, however the System Event Log also just interprets the error as a Windows error
image

The Application Event Log is where we write the actual error
image

Comment (iglendd) [reply pkg/util/winutil/servicemain/servicemain.go:148]: nice

**Review by bkabrda — APPROVED** [review]
Approved, but I'd appreciate if you could add a CODEOWNERS entry that would mark windows-agent as owners of the `tasks/windows_resources.py` file.

**Review by ogaca-dd — APPROVED** [review]

**Review by truthbk — APPROVED** [review]
Beautiful PR; fixes so many things. Looked mostly at the core and trace-agents, but overall looks ready to me! 🚀

**Review by derekwbrown — APPROVED** [review]

**Review by knusbaum — APPROVED** [review]

**Comment by clarkb7** [comment]
rebasing main for conflicts

**Pull request merged**
